Add  ARLS.go to Gonum

rejones7go opened the pull request:
Proposing to add code called ARLS.go.
The name means "Auto Regularized Linear Solver".
This package solves ill-conditioned systems of linear equations automatically.
The code is robust in that it has no algorithmic panics at all, as problematic linear systems are
considered to be its normal input.
See more about this package at http://www.rejones7.net/ArlsGo/index.htm .

Please take a look.

kortschak commented on arls.go:
```diff
@@ -0,0 +1,1199 @@
+package arls
+/*
+Copyright (c) 2020 The Go Authors. All rights reserved.
```
Is it OK to change this to the Gonum license? If it is, please amend this to be 
```
// Copyright ©2021 The Gonum Authors. All rights reserved.
// Use of this source code is governed by a BSD-style
// license that can be found in the LICENSE file.
```

kortschak reviewed (commented):
Modulo the license query (if it needs to retain the Go license, that's fine fine though) I'm happy to merge this on the basis that we'll massage it into idiomatic Go. I'd certainly like you to be involved in reviewing the changes that we make to do that if that's OK with you, Ron.

Thanks for making this contribution.

rejones7go commented:
Sure. I didn’t realize gonum had a separate license.

On Thu, Feb 25, 2021 at 9:54 PM Dan Kortschak <notifications@github.com>
wrote:

> *@kortschak* commented on this pull request.
>
> Modulo the license query (if it needs to retain the Go license, that's
> fine fine though) I'm happy to merge this on the basis that we'll massage
> it into idiomatic Go. I'd certainly like you to be involved in reviewing
> the changes that we make to do that if that's OK with you, Ron.
>
> Thanks for making this contribution.
> ------------------------------
>
> In arls.go <https://github.com/gonum/exp/pull/48#discussion_r583381839>:
>
> > @@ -0,0 +1,1199 @@
>
> +package arls
>
> +/*
>
> +Copyright (c) 2020 The Go Authors. All rights reserved.
>
>
> Is it OK to change this to the Gonum license? If it is, please amend this
> to be
>
> // Copyright ©2021 The Gonum Authors. All rights reserved.
>
> // Use of this source code is governed by a BSD-style
>
> // license that can be found in the LICENSE file.
>
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gonum/exp/pull/48#pullrequestreview-599250756>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASLYHNKMWKTMNSZFXLNTHNTTA4SPVANCNFSM4YHW23LA>
> .
>

kortschak commented:
Would you please make the license change?

kortschak reviewed (approved):
Thank you.

rejones7go commented:
Yes, will do.

Actually, I just changed the GO license to the brief 3-line GONUM license,
which references the LICENSE file.
I also did a "commit" for arls and test_arls.
Is that it? Is that all there is to do?
No new PR?

On Mon, Mar 1, 2021 at 3:46 PM Dan Kortschak <notifications@github.com>
wrote:

> Would you please make the license change?
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/gonum/exp/pull/48#issuecomment-788363514>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASLYHNMOQ2G25ZUDD323AIDTBQKMDANCNFSM4YHW23LA>
> .
>

kortschak commented:
This is all good. Thanks for that. Soon I plan to go through the code and massage it into idiomatic Go. I'll cc you on changes.

kortschak commented on arls.go:
```diff
@@ -0,0 +1,1173 @@
+    if ur >= nr {
+        x, check = rmslambda(b, &U, S, &V, ur, 0.0)
+    } else {
+        noise := Utb
+        noise.SliceVec(ur, mn)
+        sigma = myrms(noise)   //sigma := myrms(Utb[ur:mn])
```
@rejones7go Can I get a clarification of this?

`SliceVec` does not slice the receiver, it returns a slice that shared the same backing data. So these three lines are equivalent to `sigma = myrms(Utb)`. I made a change to the code so that it reflects the comment, but that then fails.

```
--- FAIL: TestArlsall (0.00s)
	    arls_test.go:15: TestArlsAll(7) failed.
```

rejones7go commented:
That's a serious error on my part.
What I need is the root mean square of the elements from index ur to mn...
I will need to investigate that.
Please show me the line you changed it to.
Ron



On Mon, Mar 1, 2021 at 11:38 PM Dan Kortschak <notifications@github.com>
wrote:

> *@kortschak* commented on this pull request.
> ------------------------------
>
> In arls.go <https://github.com/gonum/exp/pull/48#discussion_r585296972>:
>
> > +        noise := Utb
> +        noise.SliceVec(ur, mn)
> +        sigma = myrms(noise)   //sigma := myrms(Utb[ur:mn])
>
> @rejones7go <https://github.com/rejones7go> Can I get a clarification of
> this?
>
> SliceVec does not slice the receiver, it returns a slice that shared the
> same backing data. So these three lines are equivalent to sigma =
> myrms(Utb). I made a change to the code so that it reflects the comment,
> but that then fails.
>
> --- FAIL: TestArlsall (0.00s)
> 	    arls_test.go:15: TestArlsAll(7) failed.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gonum/exp/pull/48#pullrequestreview-601469036>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASLYHNORHPGLD5LRBWDUCO3TBSBVHANCNFSM4YHW23LA>
> .
>

kortschak commented:
To match the comment, the code should be
```
		noise := Utb.SliceVec(ur, mn).(*mat.VecDense)
		sigma = myrms(noise) // sigma := myrms(Utb[ur:mn])
```

rejones7go commented:
OK, thanks.
This fix is important.
The tests in test_arls() didn’t highlight the bug because all the tests are
using tiny systems.
I will commit a fix for test_arls (and arls) but it will take me a day or
two to make all the adjustments and add a “real world” problem.
Thanks a million for noticing this.
Ron

On Tue, Mar 2, 2021 at 1:21 PM Dan Kortschak <notifications@github.com>
wrote:

> To match the comment, the code should be
>
> 		noise := Utb.SliceVec(ur, mn).(*mat.VecDense)
> 		sigma = myrms(noise) // sigma := myrms(Utb[ur:mn])
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gonum/exp/pull/48#issuecomment-789187711>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASLYHNI7VL3IUFESLHJ3VHLTBVCFXANCNFSM4YHW23LA>
> .
>

kortschak commented:
No worries. Note that you'll need to pull master since the code has been gofmt'd and moved into `arls/`.

rejones7go commented:
Dan... I just updated arls and test_arls.
Now, when I reintroduce the bug you found -- just to see how the tests
behave -- instead of NONE of the tests detecting, SIX of them do,

A couple of other things I want to mention:

1. I noticed that the copyright statement is not actually correct.
You can Google "Correct form of a copyright notice" to see the recommended
form.
Basically, it is redundant to say "Copyright © 2021 by ... "
The correct form is just "Copyright 2021..... " or else  "(c) 2021 by ...."
Doesn't matter to me, but you might want to think about it.

2. Since Arls has a fairly robust and efficient general inequality
constraint solver,
it only takes a dozen lines of code each to made routines which will :
-- force a solution to be non-decreasing (i.e., casually, "rising")
-- force a solution to be "falling"
-- force a solution to be concave up (or flat)
-- force a solution to be concave down (or flat)
I have done all these in my python code recently and they worked very well.
I wonder if you would think it would be worthwhile to add these little
extras?
Of course, they could all be wrapped in to one routine, depending on Go's
style for such things.

Thanks for everything!
Ron

On Tue, Mar 2, 2021 at 3:16 PM Dan Kortschak <notifications@github.com>
wrote:

> No worries. Note that you'll need to pull master since the code has been
> gofmt'd and moved into arls/.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gonum/exp/pull/48#issuecomment-789259650>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASLYHNJUULKNHN2YR75G6ADTBVPVBANCNFSM4YHW23LA>
> .
>

kortschak commented:
Thanks, Ron.

Would you send a PR with those changes?

1. I doubt this is a significant concern since the intention is clear ('(c)' is really just a typographic workaround from the period of typesetting when '©' was not available just as underlining is a replacement for italics).

2. When @vladimir-ch has time to consider this, that would be an issue for him.

rejones7go commented:
Sorry... do you mean the ideas for rising/falling/concave etc?

On Thu, Mar 4, 2021 at 6:54 PM Dan Kortschak <notifications@github.com>
wrote:

> Thanks, Ron.
>
> Would you send a PR with those changes?
>
>    1.
>
>    I doubt this is a significant concern since the intention is clear
>    ('(c)' is really just a typographic workaround from the period of
>    typesetting when '©' was not available just as underlining is a replacement
>    for italics).
>    2.
>
>    When @vladimir-ch <https://github.com/vladimir-ch> has time to
>    consider this, that would be an issue for him.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gonum/exp/pull/48#issuecomment-791086013>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASLYHNIFBDCZKXKHYUAQRELTCA2U5ANCNFSM4YHW23LA>
> .
>

rejones7go commented:
I guess you mean for the bug fix to arls and the resulting changes to
test_arls

Ron

On Thu, Mar 4, 2021 at 6:54 PM Dan Kortschak <notifications@github.com>
wrote:

> Thanks, Ron.
>
> Would you send a PR with those changes?
>
>    1.
>
>    I doubt this is a significant concern since the intention is clear
>    ('(c)' is really just a typographic workaround from the period of
>    typesetting when '©' was not available just as underlining is a replacement
>    for italics).
>    2.
>
>    When @vladimir-ch <https://github.com/vladimir-ch> has time to
>    consider this, that would be an issue for him.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/gonum/exp/pull/48#issuecomment-791086013>, or
> unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ASLYHNIFBDCZKXKHYUAQRELTCA2U5ANCNFSM4YHW23LA>
> .
>

kortschak commented:
Yeah, the changes for the bug fix.
